File: src/types/processor-error.ts
import { NexLocation } from './api/locations';
import { DataConfirmation } from './data-confirmation';

export interface ProcessorError {
  type:
    | 'MISSING_API_KEY'
    | 'INVALID_API_KEY'
    | 'MISSING_SUBDOMAIN'
    | 'LOCATION_REQUIRED'
    | 'NO_LOCATIONS_FOUND'
    | 'MISSING_DAYS'
    | 'MISSING_APPOINTMENT_TYPE_NAME'
    | 'INTERNAL_ERROR';
  resolutionData?: ErrorResolutionData;
}

export const errorMessages: Record<ProcessorError['type'], string> = {
  MISSING_API_KEY: 'Api key is required',
  INVALID_API_KEY: 'Api key is invalid',
  MISSING_SUBDOMAIN: 'Subdomain is required',
  LOCATION_REQUIRED: 'At least one location must be selected',
  INTERNAL_ERROR: 'An unexpected error occurred',
  NO_LOCATIONS_FOUND: 'No locations found',
  MISSING_DAYS: 'Days is required',
  MISSING_APPOINTMENT_TYPE_NAME: 'Appointment type name is required',
};

type ErrorResolutionData =
  | { type: 'MESSAGE'; payload: string }
  | { type: 'LOCATIONS'; payload: NexLocation[] }
  | { type: 'CONFIRMATION'; payload: DataConfirmation }
  | { type: 'None'; payload: null };

File: src-tauri/src/services/processors/types/processor_error.rs
use serde::Serialize;

use crate::{
    api::types::locations::Location,
    services::processors::types::data_confirmation::DataConfirmation,
};

#[derive(Serialize, Debug, Clone)]
#[serde(
    tag = "type",
    content = "resolutionData",
    rename_all = "SCREAMING_SNAKE_CASE"
)]
pub enum ProcessorError {
    MissingApiKey,
    InvalidApiKey,
    MissingSubdomain,
    LocationRequired(ErrorResolutionData),
    NoLocationsFound,
    MissingDays,
    MissingAppointmentTypeName,
    NeedsConfirmation(ErrorResolutionData),
    InternalError(ErrorResolutionData),
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", content = "payload", rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ErrorResolutionData {
    Message(String),
    Locations(Vec<Location>),
    Confirmation(DataConfirmation),
    None,
}
